lib/spack/spack/solver/splices.lp

```diff
@@ -0,0 +1,56 @@
+% Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+% Spack Project Developers. See the top-level COPYRIGHT file for details.
+%
+% SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+%=============================================================================
+% These rules are conditionally loaded to handle the synthesis of spliced
+% packages.
+% =============================================================================
+% Consider the concrete spec:
+% foo@2.72%gcc@11.4 arch=linux-ubuntu22.04-icelake build_system=autotools ^bar ...
+% It will emit the following facts for reuse (below is a subset)
+% installed_hash("foo", "xxxyyy")
+% hash_attr("xxxyyy", "hash", "foo", "xxxyyy")
+% hash_attr("xxxyyy", "version", "foo", "2.72")
+% hash_attr("xxxyyy", "node_os", "ubuntu22.04")
+% hash_attr("xxxyyy", "hash", "bar", "zzzqqq")
+% hash_attr("xxxyyy", "depends_on", "foo", "bar", "link")
+% Rules that derive abi_splice_conditions_hold will be generated from 
+% use of the `can_splice` directive. The will have the following form:
+% can_splice("foo@1.0.0+a", when="@1.0.1+a", match_variants=["b"]) --->
+% abi_splice_conditions_hold(0, node(SID, "foo"), "foo", BashHash) :-
+%   installed_hash("foo", BaseHash),
+%   attr("node", node(SID, SpliceName)),
+%   attr("node_version_satisfies", node(SID, "foo"), "1.0.1"),
+%   hash_attr("hash", "node_version_satisfies", "foo", "1.0.1"),
+%   attr("variant_value", node(SID, "foo"), "a", "True"),
+%   hash_attr("hash", "variant_value", "foo", "a", "True"),
+%   attr("variant_value", node(SID, "foo"), "b", VariVar0),
+%   hash_attr("hash", "variant_value", "foo", "b", VariVar0),
```

Comment:
I'd be interested in a future revision if we could do this as a conditional rule rather than generation, but I think this is fine for now.

Reply:
Resolving because github won't let us merge without resolving all conversations.

Reply:
So for a while I had it that packages emitted `can_splice` facts, and `abi_splice_conditions_hold` was a rule that reasoned about `can_splice` facts directly. This was 200-400% slower, since the general rule was very tricky to ground (and would often cause clingo to OOM).